### ISSUE TITLE
Problem accessing system bus

### ISSUE THREAD

**Issue by jgrigalck** [sidorares/node-dbus#50]
I am unable to open the system bus for access.  When I try, I get the following error:
/home/jgriglack/Projects/ikabit/build/rootfs/share/ikabit-equipment-service/node_modules/dbus-native/lib/handshake.js:95
            if (err) throw err;
                           ^
Error: EISDIR, read
It appears that it is unable to find a certificate.

Here is part of the code:

time_dbus.js

```js
DBUS_TIME_INTERFACE = 'org.freedesktop.DBus.Properties';
DBUS_TIME_DESTINATION = 'org.freedesktop.timedate1';
DBUS_TIME_PATH = '/com/freedesktop/timedate1';
DBUS_TIME_INTROSPECT = 'org.freedesktop.DBus.Introspectable';
DBUS_TIME_PROPERTIES = 'org.freedesktop.DBus.Properties';

var _ = require('underscore');
var equipment_dbus = require('./equipment_dbus');
var Q = require('q');

//var bus = equipment_dbus.bus;

var dbus = require('dbus-native');

function time_dbus_bus()
{
    return dbus.systemBus();
}

var bus = time_dbus_bus();


/**
 * initialize will initialize the nameList and typeList variables used
 * for validating the names to Set and for sending the correct type to
 * the Set dbus function.
 */
function initialize()
{
    initializeSignals();
    bus.getService(DBUS_TIME_DESTINATION).getInterface(DBUS_TIME_PATH,
        DBUS_TIME_INTROSPECT, function(err, iface) 
    {
        if (err) 
        {
            console.log("bus.getInterface ", err);
        }
        else
        {
            var responseCallback = function(err,result)
            {
                if (err != null)
                {
                    console.log("Error! ", err);
                }

                if (result != null)
                {
                    var tmp = result.match(/property type=.*"/g);
                    var type = [];
                    var name = [];
                    for (var i = 0; i < tmp.length; i++)
                    {
                        var typeStart = tmp[i].lastIndexOf("type=")+6;
                        typeList[i] = tmp[i].substring(typeStart, typeStart+1);
                        var nameStart = tmp[i].lastIndexOf("name=")+6;
                        var nameEnd = tmp[i].lastIndexOf("access")-2;
                        nameList[i] = tmp[i].substring(nameStart, nameEnd);
                    }
                }
            }

            iface.Introspect(responseCallback);
        }
    });
}

exports.initialize = initialize;

/**
 * @brief Set the system clock
 * @param timeValue  the time value
 */
function setTime(timeValue)
{
    var deferred = Q.defer();
    bus.getService(DBUS_TIME_DESTINATION).getInterface(DBUS_TIME_PATH,
        DBUS_TIME_INTERFACE, function(err, iface)
    {
        if (err)
        {
            deferred.reject(err);
        }
        else
        {
            iface.SetTime(timeValue, 0, 0, function(err, result)
            {
                if (err)
                {
                    deferred.reject(err);
                }
                else
                {
                    deferred.resolve(result);
                }
            });
        }
    });
    return deferred.promise;
}

exports.setTime = setTime;
```
time.js

```js
var time_dbus = require('../dbus/time_dbus');
var defs = require('../definitions');
var statusEnum = defs.statusEnum;

require('shelljs/global');

//logging setup from
//http://brilliantbritz.com/setting-up-winston-logger-in-nodejs/
var winston = require('winston');

var logger = new (winston.Logger)({
 transports: [
     new (winston.transports.Console)({ level: 'error',
                                        colorize: true}),
 ]
});

module.exports = function (app)
{

/**
 * @brief get the time
 *
 * @name GET /v1/time
 */
    app.get('/v1/time', function(request, response)
    {
        function truncateDecimals (num, digits) 
        {
            var numS = num.toString();
            var decPos = numS.indexOf('.');
            var substrLength = decPos == -1 ? numS.length : 1 + decPos + digits;
            var trimmedResult = numS.substr(0, substrLength);
            var finalResult = isNaN(trimmedResult) ? 0 : trimmedResult;

            return parseFloat(finalResult);
        }

        /* The javascript Date.now() function returns a time in
         * milliseconds, so we divide by 1000 to get seconds
         */
        var timeVal = truncateDecimals((Date.now() / 1000), 0);

        response.send(statusEnum.STATUS_OK, 
                      JSON.stringify({"UTCSeconds":timeVal}));
    });

/**
 * @brief set the time
 * 
 * request body json format:
 * 
 * { "UTCSeconds": <int>, "Timezone":"America/New_York" }
 * where the UTCSeconds is the UTC Epoch seconds
 *           Timezone is one of the valid timezones in 
 *               /usr/share/zoneinfo/zone.tab
 * @name POST /v1/time
 */
    app.post('/v1/time', function(request, response) 
    {
        var setCount = 0;
        var timeRequest = request.body;
        var status = {};

        if (timeRequest.hasOwnProperty("UTCSeconds"))
        {
            // timedated wants the time in nanoseconds
            var timeValue = timeRequest.UTCSeconds * 1000000;

            if (timeValue > 0)
            {
            time_dbus.setTime(timeValue)
            .then(function(result)
            {
            setCount++;
            })
            .catch( function(err)
            {
                console.log("time interface ", err);
            });
            }
        }

        response.send(statusEnum.STATUS_OK, 
                      JSON.stringify(status));
    });

}
```

**Comment by sidorares** [comment]
The error means that filename here https://github.com/sidorares/node-dbus/blob/master/lib/handshake.js#L29 is directory and we are trying to read it as file. Could you add `console.log(filename)` after line 28 and tell what it's printing?

**Comment by sidorares** [comment]
also please add `console.log(data)` here - https://github.com/sidorares/node-dbus/blob/master/lib/handshake.js#L93

It looks that cookie context is empty string and I would expect answer for previous comment would be "~/.dbus-keyrings/"

**Comment by jgrigalck** [comment]
I added the two lines you requested, though I labelled them so I could see what they output (I hope you don’t mind).  The output was:

express deprecated app.configure: Check app.get('env') in an if statement build/rootfs/share/ikabit-equipment-service/js/server.js:355:5
Data:  [ 'REJECTED', 'EXTERNAL', 'DBUS_COOKIE_SHA1', 'ANONYMOUS\r' ]
data jg:  [ '' ]
Initializing heartrate DBUS interface
Initializing DBUS interface
Got motor dbus
Model:625
Frequency:50
filename jg:  /home/jgriglack/.dbus-keyrings/
init:  0
Warning: desired state is current state
attempt to stop motor
Listening on localhost port 3000

/home/jgriglack/Projects/ikabit/build/rootfs/share/ikabit-equipment-service/node_modules/dbus-native/lib/handshake.js:97
            if (err) throw err;
                           ^
Error: EISDIR, read

I checked, and the directory /home/jgriglack/.dbus-keyrings/ exists and contains one file named “org_freedesktop_general”.

**Comment by sidorares** [comment]
Ok, that's a bug in the client. See http://dbus.freedesktop.org/doc/dbus-specification.html#auth-mechanisms-sha : 

> Each server has a "cookie context," which is a name that identifies a set of cookies that apply to that server. A sample context might be "org_freedesktop_session_bus". Context names must be valid ASCII, nonzero length, and may not contain the characters slash ("/"), backslash ("\"), space (" "), newline ("\n"), carriage return ("\r"), tab ("\t"), or period ("."). There is a default context, "org_freedesktop_general" that's used by servers that do not specify otherwise.

We need to add default context. I'll try to fix tomorrow, but feel free to try fixing yourself - PRs welcomed :)

**Comment by sidorares** [comment]
could you try "default-context" branch?

**Comment by jgrigalck** [comment]
That appears to work.

**Comment by sidorares** [comment]
great, I'll publish it now

**Comment by sidorares** [comment]
published as v0.1.2
